src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -134,151 +144,19 @@ internal static partial class DependentTypeFinder
+        /// that match the provided <paramref name="typeMatches"/> predicate.
         /// </summary>
-        public static Task<ImmutableArray<INamedTypeSymbol>> FindTransitivelyImplementingStructuresClassesAndInterfacesAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            IImmutableSet<Project> projects,
-            CancellationToken cancellationToken)
-        {
-            return FindTypesFromCacheOrComputeAsync(
-                type, solution, projects, s_typeToTransitivelyImplementingStructuresClassesAndInterfacesMap,
-                c => FindTransitivelyImplementingStructuresClassesAndInterfacesWorkerAsync(type, solution, projects, c),
-                cancellationToken);
-        }
-
-        private static Task<ImmutableArray<INamedTypeSymbol>> FindTransitivelyImplementingStructuresClassesAndInterfacesWorkerAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            IImmutableSet<Project> projects,
-            CancellationToken cancellationToken)
-        {
-            return FindDerivedAndImplementingTypesAsync(
-                type, solution, projects, transitive: true, cancellationToken);
-        }
-
-        /// <summary>
-        /// Used for implementing the Inherited-By relation for progression.
-        /// </summary>
-        public static Task<ImmutableArray<INamedTypeSymbol>> FindImmediatelyDerivedAndImplementingTypesAsync(
-            INamedTypeSymbol type,
-            Solution solution,
-            CancellationToken cancellationToken)
-        {
-            return FindTypesFromCacheOrComputeAsync(
-                type, solution, projects: null,
-                cache: s_typeToImmediatelyDerivedAndImplementingTypesMap,
-                findAsync: c => FindDerivedAndImplementingTypesAsync(
-                    type, solution, projects: null, transitive: false, c),
-                cancellationToken: cancellationToken);
-        }
-
-        private static Task<ImmutableArray<INamedTypeSymbol>> FindDerivedAndImplementingTypesAsync(
+        /// <param name="shouldContinueSearching">Called when a new match is found to check if that type's inheritance
+        /// tree should also be walked down.  Can be used to stop the search early if a type could have no types that
+        /// inherit from it that would match this search.</param>
+        /// <param name="transitive">If this search after finding the direct inherited types that match the provided
+        /// predicate, or if the search should continue recursively using those types as the starting point.</param>
+        private static async Task<ImmutableArray<INamedTypeSymbol>> DescendInheritanceTreeAsync(
             INamedTypeSymbol type,
             Solution solution,
             IImmutableSet<Project> projects,
-            bool transitive,
-            CancellationToken cancellationToken)
```

Comment:
renamed these to 'Descend' or 'Add' to better help explain what they're doing.  'Find' was fairly generic.